lib/liblink/socket/sendmsg.ex

```diff
@@ -0,0 +1,160 @@
+  alias Liblink.Socket.Sendmsg.Fsm
+  alias Liblink.Socket.Sendmsg.Impl
+
+  require Logger
+
+  @dialyzer [:unknown]
+
+  @doc """
+  initializes the sendmsg thread.
+  """
+  @spec start() :: {:ok, pid}
+  def start() do
+    GenServer.start(__MODULE__, [])
+  end
+
+  @impl true
+  def init(_args) do
+    Impl.init()
+  end
+
+  @doc """
+  halt the sendmsg thread.
+  """
+  def halt(pid, timeout \\ 5_000) do
+    tag = Process.monitor(pid)
+    GenServer.cast(pid, :halt)
+
+    receive do
+      {:DOWN, ^tag, :process, _pid, _reason} ->
+        Process.demonitor(tag)
+        :ok
+    after
+      timeout ->
+        Process.demonitor(tag)
+        Process.exit(pid, :kill)
+        :ok
+    end
+  end
+
+  @doc false
+  def attach(pid, device, timeout \\ 5_000) do
+    GenServer.call(pid, {:attach, device}, timeout)
+  end
+
+  @doc """
+  sends a synchronous message. the server will try to meet the timeout
+  deadline and will discarde expired messages.
+
+  this function uses `GenServer.call` to perform this request and
+  catches exit signals. this should be usually fine as we wait a
+  little longer for a reply message. however, in the rare event we
+  miss the reply message the client must in this case be prepared to
+  discard garbage messages that are two-element tuples with a
+  reference at the first element. refer to `GenServer.call`
+  documentation for more information about this.
+  """
+  def sendmsg(pid, message, :infinity) do
+    GenServer.call(pid, {:sendmsg, message}, :infinity)
+  end
+
+  def sendmsg(pid, message, timeout_in_ms) when is_integer(timeout_in_ms) do
+    timeout = :erlang.convert_time_unit(timeout_in_ms, :millisecond, :native)
+    deadline = :erlang.monotonic_time() + timeout
+
+    try do
+      GenServer.call(pid, {:sendmsg, message, deadline}, timeout_in_ms + 100)
+    catch
+      :exit, {:timeout, {GenServer, :call, _}} ->
+        {:error, :timeout}
+    end
+  end
+
+  @doc """
+  send an asynchornous message. notice there are no guarantees about
+  message delivery. if you need confirmation use `sendmsg/3`.
+  """
+  def sendmsg_async(pid, message, :infinity) do
+    GenServer.cast(pid, {:sendmsg, message})
+  end
+
+  def sendmsg_async(pid, message, timeout_in_ms) when is_integer(timeout_in_ms) do
+    timeout = :erlang.convert_time_unit(timeout_in_ms, :millisecond, :native)
+    deadline = :erlang.monotonic_time() + timeout
+
+    GenServer.cast(pid, {:sendmsg, message, deadline})
+  end
+
+  @impl true
+  def handle_call(message, from, state) do
+    case message do
+      :halt ->
+        Impl.halt(:sync, state)
+
+      {:attach, device} ->
+        Impl.attach(device, :sync, state)
+
+      {:sendmsg, message} ->
+        Impl.sendmsg(message, :sync, state)
+
+      {:sendmsg, message, deadline} ->
+        Impl.sendmsg(message, deadline, :sync, state)
+
+      message ->
+        _ = Logger.warn("[socket.sendmsg] ignoring unexpected message")
+        {:noreply, state}
```

Comment:
Já que isto é um `call`, vai jogar um `:exit, {:timeout, ...}` para quem chamou ou bloquear indefinidamente, se o _timeout_ for `:infinity`, não?